### PR TITLE
Updated Big-O Notation.markdown

### PR DESCRIPTION
**Pull request by nb2998**
### Checklist

- [x] I've looked at the [contribution guidelines](https://github.com/raywenderlich/swift-algorithm-club/blob/master/.github/CONTRIBUTING.md).
- [x] This pull request is complete and ready for review.

### Description

Fixes #632. It improves the Big-O Notation.markdown by providing code snippets to better understand the topic.

### CODE REVIEW

File: Big-O Notation.markdown [hunk 1]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+**O(n!)** | factorial | **Intolerably slow.** It literally takes a million years to do anything.  
+  
 
+Some examples to better understand the Big(O) notation:
+
+**O(1)**
+
+  The most common example with O(1) complexity is accessing an array index.
+  
+  ```c++
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]: In Swift:

```swift
let value = array[5]
```

File: Big-O Notation.markdown [hunk 2]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+  for(int i=0; i<n; i *= 2){
+      cout<<i<<endl;  // instead of simply incrementing, 'i' is increased by 2 times itself in each run.
+  }
+  ```
+  Binary Search Algorithm is an example of O(log n) complexity.
+  
+  
+**O(n)**
+
+  ```c++
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
for i in stride(from: 0, to: n, by: 1) {
  print(array[i])
}
```

File: Big-O Notation.markdown [hunk 3]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+      cout<<a[i]<<endl;
+  } 
+  ```
+  
+  Array Traversal and Linear Search are examples of O(n) complexity.  
+  
+  
+**O(n log n)**
+
+  ```c++
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
for i in stride(from: 0, to: n, by: 1) {
  var j = 1
  while j < n {
    j *= 2
    // do constant time stuff
  }
}
```

or

```swift
for i in stride(from: 0, to: n, by: 1) {
  func index(after i: Int) -> Int? { // multiplies `i` by 2 until `i` >= `n`
    return i < n ? i * 2 : nil 
  }
  for j in sequence(first: 1, next: index(after:)) {
    // do constant time stuff
  }
}
```

File: Big-O Notation.markdown [hunk 4]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+  }
+```
+  
+  Merge Sort and Heap Sort are examples of O(n log n) complexity.  
+  
+  
+**O(n^2)**
+
+  ```c++
+  for(int i = 0; i < n; i++) {
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
for i  in stride(from: 0, to: n, by: 1) {
  for j in stride(from: 1, to: n, by: 1) {
    // do constant time stuff
  }
}
```

File: Big-O Notation.markdown [hunk 5]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+    }
+  }  
+  ```
+  
+  Traversing a simple 2-D array and Bubble Sort are examples of O(n^2) complexity.
+  
+  
+**O(n^3)**
+
+  ```c++
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
for i in stride(from: 0, to: n, by: 1) {
  for j in stride(from: 1, to: n, by: 1) {
    for k in stride(from: 1, to: n, by: 1) {
      // do constant time stuff
    }
  }
}
```

File: Big-O Notation.markdown [hunk 6]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+  }  
+  ```  
+  
+**O(2^n)**
+
+  Algorithms with running time O(2^N) are often recursive algorithms that solve a problem of size N by recursively solving two smaller problems of size N-1.
+  The following example prints all the moves necessary to solve the famous "Towers of Hanoi" problem for N disks.
+
+  ```c++
+  void solve_hanoi(int N, string from, string to, string spare) {
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
func solveHanoi(N: Int, from: String, to: String, spare: String) {
  guard n >= 1 else { return }
  if N > 1 {
    solveHanoi(N: N - 1, from: from, to: spare, spare: to)
  } else {
    solveHanoi(N: N-1, from: spare, to: to, spare: from)
  }
}
```

File: Big-O Notation.markdown [hunk 7]
```diff
@@ -15,8 +15,114 @@ Big-O | Name | Description
...
+  } 
+  ```
+  
+  
+**O(n!)**
+
+  The most trivial example of function that takes O(n!) time is given below.
+
+  ```c++
+    void nFacFunc(int n) {
```
Comment (kelvinlauKL) [inline Big-O Notation.markdown]:
```swift
func nFacFunc(n: Int) {
  for i in stride(from: 0, to: n, by: 1) {
    nFactFunc(n - 1)
  }
}
```

**Review by kelvinlauKL — COMMENTED** [review]
Nice work! 

Since this is a Swift repository, we'd like our examples to be in Swift. I've left a note in each code snippet you wrote with a Swift version. 

We don't get to use C-style `for` loops (this capability got removed from the language 2 years ago), but I've listed the equivalent.

**Comment by kelvinlauKL** [comment]
Just noticed this is in your repo, hehe. Don't forget to make a PR to our main repo.

**Comment by nb2998** [comment]
Done. Thanks a lot! I was trying something new and ended up creating a PR to the forked repo, sorry!

**Pull request closed without merging**